omroneipApp/src/drvOmroneip.cpp

```diff
@@ -519,1517 +577,1381 @@ asynStatus drvOmronEIP::optimiseTags()
+    epicsThreadSleep(0.1);
+  }
+  omronEIPPoller *pPoller = pollerList_.at(threadName);
+  double interval = pPoller->updateRate_;
+  for (auto tag : tagMap_)
+  {
+    if (tag.second->pollerName == threadName)
+      pPoller->myTagCount_ += 1;
+  }
+  asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Starting poller: %s with interval: %f\n", driverName, functionName, threadName.c_str(), interval);
+  auto startTime = std::chrono::system_clock::now();
+  while (!omronExiting_)
+  {
+    startTime = std::chrono::system_clock::now();
+    double waitTime = interval - ((double)timeTaken / 1E9);
+    if (waitTime >= 0)
     {
-      // Expand array and add "start:array" and "end:array" either side
-      std::vector<std::string> arrayDtypes = expandArrayRecursive(rawMap, dtype);
-      expandedRow.push_back("start:array");
-      expandedRow.insert(std::end(expandedRow), std::begin(arrayDtypes), std::end(arrayDtypes));
-      expandedRow.push_back("end:array");
+      epicsThreadSleep(waitTime);
     }
-    else if (dtype.substr(0,6)=="start:" || dtype.substr(0,4)=="end:")
+    else
     {
-      // The struct we are looking through has already been expanded and this is a tag denoting the struct, therefore we just copy it in
-      expandedRow.push_back(dtype);
+      waitTime = 0;
+      asynPrint(pasynUserSelf, ASYN_TRACE_WARNING, "%s:%s Warn, Reads taking longer than requested! %f > %f\n", driverName, functionName, ((double)timeTaken / 1E9), interval);
     }
-    else
+
+    for (auto x : tagMap_)
     {
-      // If we get here, then we assume that we have a structure, but it could be a typo
-      try
+      if (x.second->pollerName == threadName && x.second->readFlag == true)
       {
-        // Recursively call this function to get all of the embedded datatypes
-        // Before each embedded struct we add the struct name and after we add "end", this is so we know when to apply extra padding required for embedded structs
-        std::vector<std::string> embeddedDtypes = expandStructsRecursive(rawMap, dtype);
-        expandedRow.push_back("start:"+dtype);
-        expandedRow.insert(std::end(expandedRow), std::begin(embeddedDtypes), std::end(embeddedDtypes));
-        expandedRow.push_back("end:"+dtype);
+        asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Reading tag: %d with polling interval: %f seconds\n", driverName, functionName, x.second->tagIndex, interval);
+        plc_tag_read(x.second->tagIndex, 0); // read from plc, we will check status and timeouts later
+        // We sleep to split up read requests within timing interval, otherwise we can get traffic jams and missed polling intervals
+        if (pPoller->myTagCount_ > 1 && pPoller->spreadRequests_)
+        {
+          pollingDelay = (interval - 0.2 * interval) / pPoller->myTagCount_;
+          epicsThreadSleep(pollingDelay);
+        }
       }
-      catch (...)
+    }
+
+    for (auto x : tagMap_)
+    {
+      if (x.second->pollerName == threadName)
       {
```

Comment:
This is intentional. If you are optimising and have 10 tags, but only 1 needs to read the entire struct. The other 9 will have readFlag = false. This means that they dont send a read request to the PLC. However they still need to read from the cached data in libplctag. There wont be a timeout unless the 1 tag that is reading from the plc times out.

I think I could make the name of the flag better however. Maybe "sendReadRequest" instead of "readFlag".